enhancements/etcd/protecting-etcd-quorum-during-control-plane-scaling.md

```diff
@@ -0,0 +1,450 @@
+
+TODO: Consult with SREs to identify metrics they might find useful for us to export
+
+#### Example of vertically scaling a Control Plane Machine
+
+To use this mechanism to vertically scale a Control Plane, the following procedure must be carried out by some user or
+operator.
+
+1. Identify a Controle Plane Machine for replacement (eg. `my-cluster-master-0`)
+1. Determine a new name for the new Control Plane Machine that will not conflict with existing Control Plane Machines
+   (eg `my-cluster-master-3`)
+1. Take a copy of the Machine resource: `oc get machine -n openshift-machine-api my-cluster-master-0 -o yaml > my-
+   cluster-master-3.yaml`
+1. Modify the Machine YAML to update the `name` and the size of the instance (eg. changing
+   `spec.providerSpec.value.instanceType` from `c5.xlarge` to `c5.2xlarge`)
+1. Create the new Machine: `oc create -f my-cluster-master-3.yaml`
+1. Delete the old Machine: `oc delete machine -n openshift-machine-api my-cluster-master-0`
+1. Wait until the process described in [Operational flow during a resize operation](#Operational-flow-during-a-resize-operation)
+   has been completed. The end result of this is that the Machine `my-cluster-master-0` will be removed
+
+### Risks and Mitigations
+
+#### Blocking removal of Machines in environments with restricted capacity
+
+This proposal assumes that when a Machine is to be replaced, that there is additional capacity available to allow the
+new Machine to be created before the old Machine is removed. This may not be true in all environments, for example if
+there is no quota left or in bare-metal environments with limited hardware available.
+
+To ensure that we do not block users from replacing Control Plane Machines in these scenarios, we must allow a user to
+remove the etcd quorum hook without it being replaced by the etcd Operator.
+When a Machine is marked for deletion, if the hook is removed, the etcd Operator will not replace it. This will allow
+users to override the mechanism and continue with their replacement in these scenarios.
+
+In the future, if a new `ControlPlane` CRD is introduced, the etcd Operator could observe the upgrade strategy on this
+CR within the cluster and take appropriate action based on this.
+We could allow users via this CRD to signal that they do not have capcity to burst during scaling operations and in
+this case the etcd Operator can remove the hooks as appropriate.
+
+#### Other controllers may interfere with the mechanism
+
+We know that some OpenShift users leverage GitOps mechanisms to manage Machines within OpenShift.
+If these GitOps systems do not correctly handle server side changes (like additional annotations),
+then they may remove the hooks after the etcd Operator has added them.
+In this case we expect a hot loop where the etcd and GitOps operators fight to add and remove the hook.
+We must test this scenario and ensure that external systems aren't going to interfere with the mechanism.
```

Comment:
the machine types are wrong, compare https://github.com/openshift/api/blob/master/machine/v1beta1/types_machine.go#L208. The list-type must be map. Otherwise, atomic is assumed. This means that Gitops tools will replace the whole list if they just want to change one hook.

Reply:
This wasn't picked up in the initial API review, this could cause an issue to change so close the code freeze

Is there any way for a CRD to have merge by default on a list or does the merging need to be handled by the operator updating them?

Perhaps we need to implement webhooks to prevent addition/removal of more than one hook at a time, that would then match the behaviour of finalizers and hopefully prevent errors from third party controllers

For a map, would it need to be something along the lines of `map[string]string` or is there precedence for doing `map[string]struct{...}` in kube?

Reply:
I haven't tested this yet, but I think to resolve this comment we need https://github.com/openshift/api/compare/master...JoelSpeed:set-list-type-hooks?expand=1

Reply:
> this could cause an issue to change so close the code freeze

Both upstream and downstream are not very strict with considering list-type changes as breaking API change. We will even do them for types that we have for 5y as stable v1. So I would not worry about that.

> Is there any way for a CRD to have merge by default on a list or does the merging need to be handled by the operator updating them?

list-type :-)